src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArray = [{ ...state }];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      resultArray.push(
+        Object.assign(
+          { ...resultArray[resultArray.length - 1] },
+          action.extraData
+        )
+      );
+    }
+
+    if (action.type === 'removeProperties') {
+      const newState = { ...resultArray[resultArray.length - 1] };
+
+      for (const key of action.keysToRemove) {
+        if (newState[key]) {
+          delete newState[key];
+        }
+      }
+
+      resultArray.push(newState);
+    }
+
+    if (action.type === 'clear') {
+      resultArray.push({});
+    }
+  }
+
+  return resultArray.slice(1);
```

Comment:
```suggestion
  return resultArray;
```
after refactor will be returned resultArray